scenario-compiler/src/main/scala/pl/touk/nussknacker/engine/definition/clazz/ClassDefinitionSet.scala

```diff
@@ -9,11 +9,40 @@ object ClassDefinitionSet {
 }
 
 case class ClassDefinitionSet(classDefinitionsMap: Map[Class[_], ClassDefinition]) {
-  lazy val unknown = get(classOf[java.lang.Object])
+  lazy val unknown: Option[ClassDefinition] = get(classOf[java.lang.Object])
+
+  private lazy val parameterlessMethodsByDeclaringTypes: Map[DeclaringClassWithMethodName, ParameterlessMethodInfo] = {
+    def getParameterlessMethods(staticMethods: Boolean) = {
+      for {
+        (clazz, classDefinition) <- classDefinitionsMap
+        (methodName, methods)    <- if (staticMethods) classDefinition.staticMethods else classDefinition.methods
+        method                   <- methods
+        if method.signatures.exists(_.parametersToList.isEmpty)
+      } yield DeclaringClassWithMethodName(clazz, methodName) -> ParameterlessMethodInfo(isStatic = staticMethods)
+    }
+    getParameterlessMethods(staticMethods = true) ++ getParameterlessMethods(staticMethods = false)
+  }
 
   def all: Set[ClassDefinition] = classDefinitionsMap.values.toSet
 
-  def get(clazz: Class[_]): Option[ClassDefinition] =
-    classDefinitionsMap.get(clazz)
+  def get(clazz: Class[_]): Option[ClassDefinition] = classDefinitionsMap.get(clazz)
+
+  def isParameterlessMethodAllowed(targetClass: Class[_], methodName: String, staticMethodsOnly: Boolean): Boolean = {
+    extractAllTypesInHierarchy(targetClass).exists(typeToCheck =>
```

Comment:
Let's move this all types hierarchy traversal to for-comprehension above so we won't do even this step in runtime

Reply:
1. Let's add a comment
2. (Optionally) Let's replace lists in `extractAllTypesInHierarchy` with `Steam` to reduce unnecessary traversal

Reply:
fixed